basic/34-subgraph/bscdemo/README.md

```diff
@@ -0,0 +1,337 @@
+
+   ```
+   /* eslint-disable prefer-const */
+   import { ethereum } from "@graphprotocol/graph-ts";
+   import { Block } from "../generated/schema";
+   
+   export function handleBlock(block: ethereum.Block): void {
+     let entity = new Block(block.hash.toHex());
+     entity.parentHash = block.parentHash;
+     entity.unclesHash = block.unclesHash;
+     entity.author = block.author;
+     entity.stateRoot = block.stateRoot;
+     entity.transactionsRoot = block.transactionsRoot;
+     entity.receiptsRoot = block.receiptsRoot;
+     entity.number = block.number;
+     entity.gasUsed = block.gasUsed;
+     entity.gasLimit = block.gasLimit;
+     entity.timestamp = block.timestamp;
+     entity.difficulty = block.difficulty;
+     entity.totalDifficulty = block.totalDifficulty;
+     entity.size = block.size;
+     entity.save();
+   }
+   ```
+
+   此时会编辑器会提示错误，暂时不用管。
+
+6. 创建一个subgraph.yaml文件来描述subgraph
+
+   ```
+   specVersion: 0.0.2
+   description: Binance Smart Chain TestNet Blocks
+   repository: https://github.com/pancakeswap
+   schema:
+     file: ./schema.graphql
+   dataSources:
+     - kind: ethereum/contract
+       name: BSCValidatorSet
+       network: bsc
+       source:
+         address: "0x0000000000000000000000000000000000001000"
+         abi: BSCValidatorSet
+       mapping:
+         kind: ethereum/events
+         apiVersion: 0.0.4
+         language: wasm/assemblyscript
+         file: ./mappings/index.ts
+         entities:
+           - Block
+         abis:
+           - name: BSCValidatorSet
+             file: ./abis/BSCValidatorSet.json
+         blockHandlers:
+           - handler: handleBlock
+   
+   ```
+
+7. 创建一个schema.graphql描述 SubGraph 使用GraphQl查询实体。
+
+   ```
+   """
+   Binance Smart Chain TestNet blocks
+   """
+   type Block @entity {
+       "ID (hash)"
+       id: ID!
+       "Parent Hash"
+       parentHash: Bytes!
+       "Uncles Hash"
+       unclesHash: Bytes!
+       "Author"
+       author: Bytes!
+       "State Root"
+       stateRoot: Bytes!
+       "Transactions Root"
+       transactionsRoot: Bytes!
+       "Receipts Root"
+       receiptsRoot: Bytes!
+       "Number"
+       number: BigInt!
+       "Gas Used"
+       gasUsed: BigInt!
+       "Gas Limit"
+       gasLimit: BigInt!
+       "Timestamp"
+       timestamp: BigInt!
+       "Difficulty"
+       difficulty: BigInt!
+       "Total Difficulty"
+       totalDifficulty: BigInt!
+       "Size"
+       size: BigInt
+   }
+   ```
+
+   至此所有的准备完成。
+
+## 三、部署我们的graph
+
+1. 安装graph-cli
+
+   ```
+   yarn global add @graphprotocol/graph-cli
+   ```
+
+2. 安装依赖
+
+   ```
+   yarn && yarn codegen
+   ```
+
+   ![image-20211104180150595](/Users/yyhaier/Documents/%E5%85%AC%E5%8F%B8%E9%A1%B9%E7%9B%AE/%E4%BB%8E%E9%9B%B6%E5%BC%80%E5%A7%8B%E6%90%AD%E5%BB%BAGraph%20Node%E9%80%9A%E8%BF%87%20GraphQL%20%E4%BB%8E%E5%8C%BA%E5%9D%97%E9%93%BE%E6%9F%A5%E8%AF%A2%E5%A4%8D%E6%9D%82%E6%95%B0%E6%8D%AE.assets/image-20211104180150595.png)
```

Comment:
same problem, or your can use mac app iPic,to upload image to free web space, and paste the url here~